#! /bin/sh /usr/share/dpatch/dpatch-run
## 01_redirect_fix.dpatch by  <svenud@ozemail.com.au>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP:   * prevent redirect code from allowing redirect to other hosts
## DP:       (Closes: #404222)

@DPATCH@
diff -urNad un~/lib/TWiki.pm un/lib/TWiki.pm
--- un~/lib/TWiki.pm	2006-10-25 02:16:05.000000000 +0200
+++ un/lib/TWiki.pm	2006-12-21 21:31:47.000000000 +0100
@@ -720,6 +720,21 @@
 
     ASSERT($this->isa( 'TWiki')) if DEBUG;
 
+    # prevent phishing byt only allowing redirect to configured host
+    if( $url =~ m!^([^:]*://[^/]*)/*(.*)?$! ) {
+        my $host = $1;
+        #remove trailing /'s to match
+        $TWiki::cfg{DefaultUrlHost} =~ m!^([^:]*://[^/]*)/*(.*)?$!;
+        my $expected = $1;
+        unless ($host eq $expected) {
+               $url = $this->getOopsUrl( 'accessdenied',
+                                def => 'topic_access',
+                                web => $this->{web} || $TWiki::cfg{UsersWebName},
+                                topic => $this->{topic} || $TWiki::cfg{HomeTopicName},
+                                params => [ 'redirect', 'unsafe redirect to '.$url.': '.$host.' does not match DefaultUrlHost' ]);
+        }
+    }
+
     my $query = $this->{cgiQuery};
     unless( $this->{plugins}->redirectCgiQueryHandler( $query, $url ) ) {
         if ( $query && $query->param( 'noredirect' )) {
